server/src/stateManagement.mjs

```diff
@@ -79,23 +79,115 @@ addDefaultUser(config.app.defaultUserId);
+  }
+
+  //getting user, group and appId from userId
+  if (userId.includes("~")){
+    user = userId.split("~")[0];
+    if (userId.includes("#")){
+      appId = userId.split("#")[1];
+      group = "~"+userId.split("#")[0].split('~')[1];
+    }
+    else{
+      group = "~"+userId.split('~')[1];
+    }
+  }
+  else if (userId.includes("#")){
+    user = userId.split("#")[0];
+    appId = userId.split("#")[1];
+  }
+  else{
+    user = userId;
+  }
+
+  //iterating over list of users in state to ensure duplicate user/appId
+  for(var key in users){
+    if (users[key].includes("~")){
+      if (user && users[key].split("~")[0]==user){
+          logger.info(`Cannot add user ${userId} as user ${user} already exists`)
+          return {isSuccess: false, msg : `Cannot add user ${userId} as user ${user} already exists`}
+      }
+      else if(appId && users[key].includes("#") && users[key].split("#")[1]==appId){
+        logger.info(`Cannot add user ${userId} as appId ${appId} already exists`)
+        return {isSuccess: false, msg : `Cannot add user ${userId} as appId ${appId} already exists`}
+      }
+    }
+    else if (users[key].includes("#")){
+      if (user && users[key].split("#")[0]==user){
+        logger.info(`Cannot add user ${userId} as appId ${user} already exists`)
+        return {isSuccess: false, msg : `Cannot add user ${userId} as appId ${user} already exists`}
+      }
+      else if (appId && users[key].split("#")[1]==appId){
+        logger.info(`Cannot add user ${userId} as user ${appId} already exists`)
+        return {isSuccess: false, msg : `Cannot add user ${userId} as user ${appId} already exists`}
+      }
+    }
+    else{
+      if (user && users[key] == user){
+        logger.info(`Cannot add user ${userId} as user ${user} already exists`)
+        return {isSuccess: false, msg : `Cannot add user ${userId} as user ${user} already exists`}
+      }
+    }
+  }
+
   state[''+userId] = JSON.parse(JSON.stringify(perUserStartState));  // Deep copy
+  if (!(group in state)){
+    state[''+group] = JSON.parse(JSON.stringify(perUserStartState)); // Deep copy
+  };
+  return {isSuccess:true, msg:""}
 }
 
 function addDefaultUser(userId) {
   addUser(userId);
   state[''+userId].isDefaultUserState = true;
 }
 
+//returns userId with given user/appId if present
+function getUserId(userId){
+  var users = Object.keys(state);
+
+  // //Checking if user or appId is present in state object
+  if (!(userId in state)){
+    for(var key in users){
+      if (users[key].includes("~")){
+        if (users[key].split("~")[0]==userId){
+          userId = users[key];
+        }
+        else if (users[key].includes("#") && users[key].split("#")[1]==userId){
+          userId = users[key];
+        }
+      }
+      else if (users[key].includes("#")){
+        if (users[key].split("#")[0]==userId || users[key].split("#")[1]==userId){
+          userId = users[key];
+        }
+      }
+    }
+  }
+  return userId;
```

Comment:
We're not accounting for a scenario in which the userId does not exist anywhere in the state

Reply:
Taken care